lib/hammer_cli/abstract.rb

```diff
@@ -233,10 +232,16 @@ def self.option(switches, type, description, opts = {}, &block)
     end
 
     def all_options
-      self.class.recognised_options.inject({}) do |h, opt|
-        h[opt.attribute_name] = send(opt.read_method)
-        h
+      @all_options ||= self.class.recognised_options.inject({}) do |hash, opt|
+        hash[opt.attribute_name] = send(opt.read_method)
+        hash[opt.attribute_name] = add_custom_defaults(opt.attribute_name) if hash[opt.attribute_name].nil?
+        hash
       end
+      @all_options
+    end
+
+    def add_custom_defaults(attr)
```

Comment:
I'm still missing the logging of the defaults being used. The `add_custom_defaults` is good candidate for doing that:
```ruby
    def add_custom_defaults(attr)
      if context[:defaults]
        value = context[:defaults].get_defaults(attr)
        logger.info("Custom default value #{value} was used for attribute #{attr}") if value
      end
      value
    end
```
I'd also make it `private`, as it is not intended to be part of public command interface